src/lib/ffi/ffi_cert.cpp

```diff
@@ -163,6 +169,79 @@ int botan_x509_cert_allowed_usage(botan_x509_cert_t cert, unsigned int key_usage
 #endif
 }
 
+int botan_x509_get_basic_constraints(botan_x509_cert_t cert, int* is_ca, size_t* limit) {
+   if(!is_ca || !limit) {
+      return BOTAN_FFI_ERROR_NULL_POINTER;
+   }
+
+#if defined(BOTAN_HAS_X509_CERTIFICATES)
+   return BOTAN_FFI_VISIT(cert, [=](const auto& c) -> int {
+      if(c.is_CA_cert()) {
+         *is_ca = 1;
+         *limit = c.path_limit();
+      } else {
+         *is_ca = 0;
+         *limit = 0;
+      }
+      return BOTAN_FFI_SUCCESS;
+   });
+#else
+   BOTAN_UNUSED(cert)
+   return BOTAN_FFI_ERROR_NOT_IMPLEMENTED;
+#endif
+}
+
+int botan_x509_get_key_constraints(botan_x509_cert_t cert, uint32_t* usage) {
+   if(!usage) {
+      return BOTAN_FFI_ERROR_NULL_POINTER;
+   }
+
+#if defined(BOTAN_HAS_X509_CERTIFICATES)
+   return BOTAN_FFI_VISIT(cert, [=](const auto& c) -> int {
+      *usage = c.constraints().value();
+      return BOTAN_FFI_SUCCESS;
+   });
+#else
+   BOTAN_UNUSED(cert)
+   return BOTAN_FFI_ERROR_NOT_IMPLEMENTED;
+#endif
+}
+
+int botan_x509_get_ocsp_responder(botan_x509_cert_t cert, botan_view_ctx ctx, botan_view_str_fn view) {
+#if defined(BOTAN_HAS_X509_CERTIFICATES)
+   return BOTAN_FFI_VISIT(cert,
+                          [=](const auto& c) -> int { return invoke_view_callback(view, ctx, c.ocsp_responder()); });
+#else
+   BOTAN_UNUSED(cert, ctx, view)
+   return BOTAN_FFI_ERROR_NOT_IMPLEMENTED;
+#endif
+}
+
+int botan_x509_is_self_signed(botan_x509_cert_t cert, int* out) {
+   if(!out) {
+      return BOTAN_FFI_ERROR_NULL_POINTER;
+   }
+
+#if defined(BOTAN_HAS_X509_CERTIFICATES)
+   return BOTAN_FFI_VISIT(cert, [=](const auto& c) {
+      if(c.is_self_signed()) {
+         *out = 1;
+      } else {
+         *out = 0;
+      }
+   });
+#else
+   BOTAN_UNUSED(cert, out)
+   return BOTAN_FFI_ERROR_NOT_IMPLEMENTED;
+#endif
+}
+
+// TODO name constraints
+// TODO extended key usage
+// TODO cert policies
+// TODO ca issuers
+// TODO crl dist points
```

Comment:
These return `vec<T>`s, which I'm not really sure how to pass through the FFI. Strings and OIDs might be possible, but `name_constraints` consists of `vec<GeneralSubtree>`, which I don't see happening at all.
That said, I don't know if we actually *need* all of those in the FFI.

Reply:
Fine to expose only a subset here, especially for the first pass